src/main/java/com/amazon/opendistroforelasticsearch/sql/parser/subquery/InSubqueryRewriter.java

```diff
@@ -0,0 +1,146 @@
+ */
+public class InSubqueryRewriter implements SubqueryRewriter {
+    private final SQLInSubQueryExpr inSubQueryExpr;
+    private final MySqlSelectQueryBlock inSubQuery;
+    private final SQLExpr joinLeftExpr;
+    private final SQLExpr joinRightExpr;
+
+    public InSubqueryRewriter(Subquery subquery) {
+        this.inSubQueryExpr = (SQLInSubQueryExpr) subquery.getSubQueryExpr();
+        this.inSubQuery = (MySqlSelectQueryBlock) inSubQueryExpr.getSubQuery().getQuery();
+        joinLeftExpr = inSubQueryExpr.getExpr();
+        joinRightExpr = rightExpr(inSubQuery);
+    }
+
+    /**
+     * Rewrite the subquery with JOIN clause. This API is NOT idempotent.
+     *
+     * @param query Subquery will be written.
+     */
+    @Override
+    public void rewrite(MySqlSelectQueryBlock query) {
+        query.setFrom(buildFrom(query));
+        query.setWhere(buildWhere(query));
+    }
+
+    /**
+     * Build From clause from input query.
+     *
+     * With the input query.
+     *       Query
+     *     /   |   \
+     * SELECT FROM     WHERE
+     *    |    |     /   |  \
+     *    *    A  c>10 AND INSubquery
+     *                      /    \
+     *                     a    Query
+     *                          /    \
+     *                       SELECT FORM
+     *                         |     |
+     *                         b     B
+     *
+     * The FROM logic should be
+     *     JOIN
+     *     / | \
+     *    A  B A.a == B.b
+     */
+    private SQLJoinTableSource buildFrom(MySqlSelectQueryBlock rootQuery) {
+        SQLJoinTableSource sqlJoin = new SQLJoinTableSource();
+        sqlJoin.setLeft(rootQuery.getFrom());
+        sqlJoin.setRight(inSubQuery.getFrom());
+        sqlJoin.setCondition(new SQLBinaryOpExpr(joinLeftExpr, SQLBinaryOperator.Equality, joinRightExpr));
+        sqlJoin.setJoinType(JoinType.JOIN);
+
+        return sqlJoin;
+    }
+
+    /**
+     * Build Where clause from input query.
+     *
+     * With the input query.
+     *       Query
+     *     /   |   \
+     * SELECT FROM     WHERE
+     *    |    |     /   |  \
+     *    *    A  c>10 AND INSubquery
+     *                      /    \
+     *                     a    Query
+     *                          /    \
+     *                       SELECT FORM
+     *                         |     |
+     *                         b     B
+     *
+     *
+     *     WHERE
+     *     /  |   \
+     *   c>10 AND B.b is NOT NULL
+     */
+    private SQLExpr buildWhere(MySqlSelectQueryBlock rootQuery) {
+        return rewriteSubquery(rootQuery.getWhere());
+    }
+
+    private SQLExpr rewriteSubquery(SQLExpr expr) {
+        if (expr instanceof SQLInSubQueryExpr) {
+            return createSubqueryReplacementCondition();
+        }
+        else if (expr instanceof SQLBinaryOpExpr) {
+            ((SQLBinaryOpExpr) expr).setLeft(rewriteSubquery(((SQLBinaryOpExpr) expr).getLeft()));
```

Comment:
store `(SQLBinaryOpExpr) expr` in a variable?

Reply:
I don't like the fact that we keep modifying methods' arguments instead of returning the result. It might not be obvious or might not even cause issues now, but it will be difficult later to track down where things have changed. Is the copy operation of the SQLBinaryOpExpr expensive?

Reply:
Understood. In case there are several vistor follow the same pattern. I think it should be fine to keep this.